### PR TITLE
fix(core): ensure change detection runs in a reasonable timeframe wit…

### LINKED ISSUE #39854: ngZoneEventCoalescing, a CanActivate guard and pace-js together break Angular's change detection
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core or zone.js

### Is this a regression?

 Yes, the previous version in which this bug was not present was: idk

### Description

 A clear and concise description of the problem...

## 🔬 Minimal Reproduction

https://github.com/tonivj5/ng-bug-can-activate-event-coalescing-and-pace
- npm i
- npm run start
- open two tabs: 1 with the angular app, other with whatever you want
- See that app works okay, then reload and quickly change tab. If you do it with good timing (I scroll tabs), app hasn't been rendered correctly (it logs less ticks than it should)
Here there's a gif with the process:
bug-can-activate-with-pace

The `true` log represents the CanActivate guard return.

As you can see, it's a strange bug :laughing:... It happens under these conditions:
	- You're not in the tab, other way of reproduce it is if you are in other tab and Angular cli reload your app.
	- You're using pace-js
	- You're using a CanActive guard, and its return an async observable (in my prod app, which is more complex, I have been able of reproduce it using `return true` into the guard)
	- And you have enabled `ngZoneEventCoalescing`.

Without one of these conditions, app works like a charm.

As you can see [here](https://github.com/tonivj5/ng-bug-can-activate-event-coalescing-and-pace/blob/8bb5844efaa0ea468a89836068382254424e4bc6/src/app/app.component.ts#L9-L13), it doesn't work anymore when app breaks. But when app works it log a tick correctly

## 🔥 Exception or Error
none

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 11.0.2
Node: 14.14.0
OS: linux x64

Angular: 11.0.2
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.2
@angular-devkit/build-angular   0.1100.2
@angular-devkit/core            11.0.2
@angular-devkit/schematics      11.0.2
@schematics/angular             11.0.2
@schematics/update              0.1100.2
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Anything else relevant?**

It happens in chrome and firefox, but in firefox is more complex to do it

### LINKED ISSUE #44314: [NgZone] Enabling bootstrap option ngZoneRunCoalescing prevents app from loading when opened in a new tab
### Which @angular/* package(s) are the source of the bug?

Zone

### Is this a regression?

Yes

### Description

After enabling the Angular option `ngZoneRunCoalescing`, all change detection is stopped when the tab is not visible since the NgZone option `shouldCoalesceRunChangeDetection` relies on `requestAnimationFrame` to trigger the next change detection. 

> ...requestAnimationFrame() calls are paused in most browsers when running in background tabs or hidden iframe...
>
> [Window.requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame)

As a result, components that rely on change detection to begin or complete work will be blocked until the page becomes visible. In this case, users who try to open an Angular app with long load time in a separate window/tab to let it load in the background will potentially make no progress on the load.

Here is a minimal app that shows the problem [run SimpleApp](https://angular-ivy-2selqj.stackblitz.io/) - [SimpleApp source](https://stackblitz.com/edit/angular-ivy-2selqj?file=src%2Fmain.ts). If you open this in a new tab, the count doesn't begin until the tab becomes visible, but if you turn off `ngZoneRunCoalescing` and open in a new tab, the count starts and continues through the end if you leave the tab hidden long enough.

I am working on an app that reads data from a potentially slow external connection, then creates a bunch of components that each further process and visualize the received data, potentially creating more components. These components rely on the Angular lifecycle to receive input bindings and update several times through services, sometimes relying on the output of other components to further process the data before it can be rendered. This is all blocked until the tab becomes visible.

Without `ngZoneRunCoalescing`, users can open the app in a separate tab and allow it to load in the background while doing other work, but with this turned on, the app sits there and doesn't start loading until the tab becomes visible.

I really want to use this option as I'm seeing a 10-30% improvement in load performance with it turned on (as long as the page is visible during the entire load)

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-2selqj

### Please provide the exception or error you saw

```true
App load blocked due to all change detection stopped
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: no
Node: v14.15.4
Package Manager: npm 6.14.10
OS: Windows 10

Angular: 12.2.11
@angular/animations: 12.2.11,
@angular/cdk: 12.2.11,
@angular/common: 12.2.11,
@angular/core: 12.2.11,
@angular/elements: 12.2.11,
@angular/forms: 12.2.11,
@angular/material: 12.2.11,
@angular/platform-browser: 12.2.11,
@angular/platform-browser-dynamic: 12.2.11,
@angular/platform-server: 12.2.11,
@angular/router: 12.2.11,
@angular/upgrade: 12.2.11,
typescript: 4.3.5,
```

### Anything else?

_No response_

### LINKED ISSUE #54544: viewTransition not working with eventCoalescing
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

Yes

### Description

Navigation "freezes" with `withViewTransitions()` and `provideZoneChangeDetection({eventCoalescing: true})`.

**Screen record:**
[screen-capture (8).webm](https://github.com/angular/angular/assets/50884231/48ba62fc-9abf-4d18-bdaf-519473274236)

### Please provide a link to a minimal reproduction of the bug

https://github.com/kkachniarz220/angular-view-transition-bug-reproduction

### Please provide the exception or error you saw

```true
No exceptions
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.2.0
Node: 20.10.0
Package Manager: npm 10.2.3
OS: win32 x64

@angular-devkit/architect       0.1702.0
@angular-devkit/build-angular   17.2.0
@angular-devkit/core            17.2.0
@angular-devkit/schematics      17.2.0
@angular/cli                    17.2.0
@schematics/angular             17.2.0
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.4
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by atscott**
…h zone coalescing

Zone event and run coalescing now races `requestAnimationFrame` and `setTimeout`. There are tradeoffs to both functions and racing the two gives us the benefits of both. This is explained in more detail in the jsdoc comment. This change also aligns the timing of zone coalescing with what will be used for zoneless.

BREAKING CHANGE: The exact timing of change detection execution when using event or run coalescing with `NgZone` is now the first of either `setTimeout` or `requestAnimationFrame`. Code which relies on this timing (usually by accident) will need to be adjusted. If a callback needs to execute after change detection, we recommend `afterNextRender` instead of something like `setTimeout`.

fixes #54544
fixes #44314
fixes #39854 (unverified but seems likely)

### CODE REVIEW

**Comment by atscott** [comment]
[Green TGP]https://fusion2.corp.google.com/presubmit/609455884/OCL:609455884:BASE:609758268:1708710829821:901fbdd5/targets)

**Review by alxhub — APPROVED** [review]

File: packages/core/src/util/raf.ts [hunk 1]
```diff
@@ -8,42 +8,71 @@
 
 import {global} from './global';
 
-export function getNativeRequestAnimationFrame() {
+/**
```
Comment (pkozlowski-opensource) [inline packages/core/src/util/raf.ts:11]: I think we should move this code to a different file or rename this one - the current `util/raf.ts` is suggesting that this is just an rAf util but now we deal with the whole scheduler.

File: packages/core/test/zone/ng_zone_spec.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import {fakeAsync, flushMicrotasks, inject, waitForAsync} from '@angular/core/te
 import {Log} from '@angular/core/testing/src/testing_internal';
 
 import {global} from '../../src/util/global';
-import {getNativeRequestAnimationFrame} from '../../src/util/raf';
+import {getCallbackScheduler} from '../../src/util/raf';
```
Comment (pkozlowski-opensource) [inline packages/core/test/zone/ng_zone_spec.ts:14]: We should add tests for the new behaviour...

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
Left a couple of comments that I would love to see addressed before merging this one.

File: packages/core/test/zone/ng_zone_spec.ts (hunk 2 above)
Comment (atscott) [reply packages/core/test/zone/ng_zone_spec.ts:14]: I'm not sure how exactly you would observe the new behavior consistently. We can patch or spy on `rAF` and `setTimeout` but that turns into a change detector test pretty easily.

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by atscott** [comment]
This PR was merged into the repository by commit 0dd441d08fb3249916f5ace7786cadf4afe49726.

**Pull request closed without merging**